src/client/client.rs

```diff
@@ -538,6 +561,266 @@ impl Client {
+                QoS::AtMostOnce => QosPid::AtMostOnce,
+                QoS::AtLeastOnce => QosPid::AtLeastOnce(self.alloc_write_pid()?),
+                QoS::ExactlyOnce => panic!("Not reached"),
+            },
+            retain: p.retain(),
+            topic_name: p.topic().to_owned(),
+            payload: p.payload().to_owned(),
+        });
+        match qos {
+            QoS::AtMostOnce => {
+                let res = timeout(self.options.operation_timeout,
+                                  self.write_only_packet(&p2)).await;
+                if let Err(Elapsed { .. }) = res {
+                    return Err(format!("Timeout writing publish after {}ms",
+                                       self.options.operation_timeout.as_millis()).into());
+                }
+                res.expect("No timeout")?;
+            }
+            QoS::AtLeastOnce => {
+                let res = timeout(self.options.operation_timeout,
+                                  self.write_response_packet(&p2)).await;
+                if let Err(Elapsed { .. }) = res {
+                    // We report this but can't really deal with it properly.
+                    // The protocol says we can't re-use the packet ID so we have to leak it
+                    // and potentially run out of packet IDs.
+                    return Err(format!("Timeout waiting for Puback after {}ms",
+                                       self.options.operation_timeout.as_millis()).into());
+                }
+                let res = res.expect("No timeout")?;
+                match res {
+                    Packet::Puback(pid) => self.free_write_pid(pid)?,
+                    _ => error!("Bad packet response for publish: {:#?}", res),
+                }
+            },
+            QoS::ExactlyOnce => panic!("Not reached"),
+        };
+        Ok(())
+    }
+
+    /// Subscribe to some topics.`read_subscriptions` will return
+    /// data for them.
+    pub async fn subscribe(&mut self, s: Subscribe) -> Result<SubscribeResult> {
+        let pid = self.alloc_write_pid()?;
+        // TODO: Support subscribe to qos == ExactlyOnce.
+        if s.topics().iter().any(|t| t.qos == QoS::ExactlyOnce) {
+            return Err("Qos::ExactlyOnce is not supported right now".into())
+        }
+        let p = Packet::Subscribe(mqttrs::Subscribe {
+            pid,
+            topics: s.topics().to_owned(),
+        });
+        let res = timeout(self.options.operation_timeout, self.write_response_packet(&p)).await;
+        if let Err(Elapsed { .. }) = res {
+            // We report this but can't really deal with it properly.
+            // The protocol says we can't re-use the packet ID so we have to leak it
+            // and potentially run out of packet IDs.
+            return Err(format!("Timeout waiting for Suback after {}ms",
+                               self.options.operation_timeout.as_millis()).into());
+        }
+        let res = res.expect("No timeout")?;
```

Comment:
Even if this should not occur due to the previous return. I think it would make more sense to use a `?` here.
so basically:

```Rust
let res = res??;
```

Maybe this is actually easier to follow:

```Rust
 let res = timeout(self.options.operation_timeout, self.write_response_packet(&p))
   .await
   .map_err(|Elapsed { .. }| {
    // Timeout occured.
    // We report this but can't really deal with it properly.
    // The protocol says we can't re-use the packet ID so we have to leak it
    // and potentially run out of packet IDs.
    Err(format!("Timeout waiting for Suback after {}ms",
           self.options.operation_timeout.as_millis()).into())
   })?;
```